File: app/src/main/java/io/treehouses/remote/Fragments/HomeFragment.java
package io.treehouses.remote.Fragments;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.bluetooth.BluetoothAdapter;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.text.TextUtils;
import android.util.Log;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ExpandableListView;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Arrays;
import java.util.List;

import io.treehouses.remote.BuildConfig;
import io.treehouses.remote.Constants;
import io.treehouses.remote.Fragments.DialogFragments.RPIDialogFragment;
import io.treehouses.remote.InitialActivity;
import io.treehouses.remote.MainApplication;
import io.treehouses.remote.Network.BluetoothChatService;
import io.treehouses.remote.R;
import io.treehouses.remote.adapter.ProfilesListAdapter;
import io.treehouses.remote.bases.BaseHomeFragment;
import io.treehouses.remote.callback.SetDisconnect;

import io.treehouses.remote.callback.NotificationCallback;
import io.treehouses.remote.pojo.NetworkProfile;
import io.treehouses.remote.utils.SaveUtils;

import static io.treehouses.remote.Constants.REQUEST_ENABLE_BT;


public class HomeFragment extends BaseHomeFragment implements SetDisconnect {
    public static final String[] group_labels = {"WiFi", "Hotspot", "Bridge", "Default"};

    private NotificationCallback notificationListener;

    private ProgressDialog progressDialog;

    private ExpandableListView network_profiles;

    private BluetoothAdapter mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
    private BluetoothChatService mChatService = null;
    private Button connectRpi, getStarted, testConnection;
    private Boolean connectionState = false;
    private Boolean result = false;
    private TextView welcome_text;
    private ImageView background, logo, internetstatus;
    private AlertDialog testConnectionDialog;
    private int selected_LED;
    private boolean checkVersionSent = false;

    private String network_ssid = "";
    private FrameLayout layout;

    private NetworkProfile networkProfile;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.activity_home_fragment, container, false);
        mChatService = listener.getChatService();
        connectRpi = view.findViewById(R.id.btn_connect);
        getStarted = view.findViewById(R.id.btn_getStarted);
        preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        testConnection = view.findViewById(R.id.test_connection);
        welcome_text = view.findViewById(R.id.welcome_home);
        background = view.findViewById(R.id.background_home);
        network_profiles = view.findViewById(R.id.network_profiles);
        logo = view.findViewById(R.id.logo_home);
        layout = view.findViewById(R.id.layout_back);
        internetstatus = view.findViewById(R.id.internetstatus);
        setupProfiles();
        showDialogOnce(preferences);
        checkConnectionState();
        connectRpiListener();
        getStarted.setOnClickListener(v -> InitialActivity.getInstance().openCallFragment(new AboutFragment()));
        testConnectionListener();

        try {
            String s = "{\"available\":[\"couchdb\",\"kolibri\",\"mariadb\",\"mastodon\"],\"installed\":[\"couchdb\",\"kolibri\",\"mastodon\"],\"running\":[\"couchdb\",\"mastodon\"],\"icon\":{\"couchdb\":\"<svg>\",\"kolibri\":\"<svg>\",\"mariadb\":\"<svg>\",\"mastodon\":\"<svg>\"},\"info\":{\"couchdb\":\"<info>\",\"kolibri\":\"<info>\",\"mariadb\":\"<info>\",\"mastodon\":\"<info>\"},\"autorun\":{\"couchdb\":\"true\",\"kolibri\":\"true\",\"mariadb\":\"true\",\"mastodon\":\"true\"}}";

            JSONObject jsonObject = new JSONObject(s);

            Log.d("HOME", jsonObject.toString());

        } catch (JSONException e) {
            e.printStackTrace();
        }
        return view;
    }

    private void setupProfiles() {

        ProfilesListAdapter profileAdapter = new ProfilesListAdapter(getContext(),
                Arrays.asList(group_labels), SaveUtils.getProfiles(getContext()));

        network_profiles.setAdapter(profileAdapter);
        network_profiles.setOnChildClickListener(new ExpandableListView.OnChildClickListener() {
            @Override
            public boolean onChildClick(ExpandableListView parent, View v, int groupPosition, int childPosition, long id) {
                if (groupPosition == 3) {
                    listener.sendMessage("treehouses default network");
                    Toast.makeText(getContext(),"Switched to Default Network", Toast.LENGTH_LONG).show();
                }
                else if (SaveUtils.getProfiles(getContext()).size() > 0 && SaveUtils.getProfiles(getContext()).get(Arrays.asList(group_labels).get(groupPosition)).size() > 0) {
                    networkProfile = SaveUtils.getProfiles(getContext()).get(Arrays.asList(group_labels).get(groupPosition)).get(childPosition);
                    listener.sendMessage("treehouses default network \n");
                    Toast.makeText(getContext(), "Configuring...", Toast.LENGTH_LONG).show();
                }
                return false;
            }
        });
    }

    private void switchProfile(NetworkProfile networkProfile) {
        if (networkProfile == null) return;
        progressDialog = ProgressDialog.show(getContext(), "Connecting...", "Switching to " + networkProfile.ssid, true);
        progressDialog.show();

        if (networkProfile.isWifi()) {
            //WIFI
            listener.sendMessage(String.format("treehouses wifi %s %s", networkProfile.ssid, networkProfile.password));
            network_ssid = networkProfile.ssid;
        } else if (networkProfile.isHotspot()) {
            //Hotspot
            if (networkProfile.password.isEmpty()) {
                listener.sendMessage("treehouses ap " + networkProfile.option + " " + networkProfile.ssid);
            } else {
                listener.sendMessage("treehouses ap " + networkProfile.option + " " + networkProfile.ssid + " " + networkProfile.password);
            }
            network_ssid = networkProfile.ssid;
        } else if (networkProfile.isBridge()) {
            //Bridge
            String temp = "treehouses bridge " + networkProfile.ssid + " " + networkProfile.hotspot_ssid + " ";
            String overallMessage = TextUtils.isEmpty(networkProfile.password) ? temp + " " : temp + " " + networkProfile.password;

            if (!TextUtils.isEmpty(networkProfile.hotspot_password)) overallMessage += " " + networkProfile.hotspot_password + " ";
            listener.sendMessage(overallMessage);
        } else {
            Log.e("Home", "UNKNOWN TYPE");
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (MainApplication.showLogDialog) {
            rate(preferences);
            showLogDialog(preferences);
        }
    }


    public void connectRpiListener() {
        connectRpi.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (connectionState) {
                    RPIDialogFragment.getInstance().bluetoothCheck("unregister");
                    mChatService.stop();
                    connectionState = false;
                    checkConnectionState();
                    return;
                }

                if (mBluetoothAdapter.getState() == BluetoothAdapter.STATE_OFF) {
                    Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
                    startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
                    Toast.makeText(getContext(), "Bluetooth is disabled", Toast.LENGTH_LONG).show();
                } else if (mBluetoothAdapter.getState() == BluetoothAdapter.STATE_ON) {
                    showRPIDialog(HomeFragment.this);
                }
            }
        });
    }

    public void testConnectionListener() {
        testConnection.setOnClickListener(v -> {
            String preference = androidx.preference.PreferenceManager.getDefaultSharedPreferences(getContext()).getString("led_pattern", "LED Dance");
            List<String> options = Arrays.asList(getResources().getStringArray(R.array.led_options));
            String[] options_code = getResources().getStringArray(R.array.led_options_commands);
            selected_LED = options.indexOf(preference);
            writeToRPI(options_code[selected_LED]);
            testConnectionDialog = showTestConnectionDialog(false, "Testing Connection...", R.string.test_connection_message, selected_LED);
            testConnectionDialog.show();
            result = false;
        });
    }

    public void checkConnectionState() {
        mChatService = listener.getChatService();
        if (mChatService.getState() == Constants.STATE_CONNECTED) {
            showLogDialog(preferences);
            transitionOnConnected();
            connectionState = true;
            checkVersionSent = true;
            writeToRPI("treehouses remote version " + BuildConfig.VERSION_CODE + "\n");

        } else {
            transitionDisconnected();
            connectionState = false;
        }
        mChatService.updateHandler(mHandler);
    }

    private void transitionOnConnected() {
        welcome_text.setVisibility(View.GONE);
        testConnection.setVisibility(View.VISIBLE);
        connectRpi.setText("Disconnect");
        connectRpi.setBackgroundResource(R.drawable.disconnect_rpi);
        background.animate().translationY(150);
        connectRpi.animate().translationY(110);
        getStarted.animate().translationY(70);
        testConnection.setVisibility(View.VISIBLE);
        layout.setVisibility(View.VISIBLE);
        logo.setVisibility(View.GONE);
    }

    private void transitionDisconnected() {
        connectRpi.setText("Connect to RPI");
        testConnection.setVisibility(View.GONE);
        welcome_text.setVisibility(View.VISIBLE);
        background.animate().translationY(0);
        connectRpi.animate().translationY(0);
        getStarted.animate().translationY(0);
        connectRpi.setBackgroundResource(R.drawable.connect_to_rpi);
        logo.setVisibility(View.VISIBLE);
        layout.setVisibility(View.GONE);
    }

    private void dismissTestConnection() {
        if (testConnectionDialog != null) {
            testConnectionDialog.cancel();
            showTestConnectionDialog(true, "Process Finished", R.string.test_finished, selected_LED);
        }
    }

    private void writeToRPI(String ping) {
        mChatService.write(ping.getBytes());
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            notificationListener = (NotificationCallback) getContext();
        } catch (ClassCastException e) {
            throw new ClassCastException("Activity must implement NotificationListener");
        }
    }

    private void dismissPDialog() { if (progressDialog != null) progressDialog.dismiss(); }

    private void checkVersion(String output) {
        checkVersionSent = false;
        if (output.contains("Usage") || output.contains("command")) {
            //CLI Needs Upgrade
            showUpgradeCLI();
        }
        else if(BuildConfig.VERSION_CODE == 2 || output.contains("true")) {
            writeToRPI("treehouses remote status\n");
            writeToRPI("treehouses upgrade --check\n");
        }
        else {
            AlertDialog alertDialog = new AlertDialog.Builder(getContext())
                    .setTitle("Update Required")
                    .setMessage("Please update Treehouses Remote, as it does not meet the required version on the Treehouses CLI.")
                    .setPositiveButton("Update", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            final String appPackageName = getActivity().getPackageName(); // getPackageName() from Context or Activity object
                            try {
                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
                            } catch (android.content.ActivityNotFoundException anfe) {
                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
                            }
                        }
                    }).create();
            alertDialog.show();
        }
    }

    private void readMessage(String output) {
        if (checkVersionSent) {
            checkVersion(output);
        } else if (output.contains(" ") && output.split(" ").length == 5) {
            checkImageInfo(output.split(" "), mChatService.getConnectedDeviceName(), internetstatus);
        } else if (matchResult(output, "true", "false") && output.length() < 14) {
            notificationListener.setNotification(output.contains("true"));
        } else if (matchResult(output, "connected", "pirateship")) {
            Toast.makeText(getContext(), "Switched to " + network_ssid, Toast.LENGTH_LONG).show();
            dismissPDialog();
        } else if (output.toLowerCase().contains("bridge has been built")) {
            dismissPDialog();
            Toast.makeText(getContext(), "Bridge Has Been Built", Toast.LENGTH_LONG).show();
        } else if (output.toLowerCase().contains("default")) {
            switchProfile(networkProfile);
        } else if (output.toLowerCase().contains("error")) {
            dismissPDialog();
            Toast.makeText(getContext(), "Network Not Found", Toast.LENGTH_LONG).show();
        } else if (!result) {
            //Test Connection
            result = true;
            dismissTestConnection();
        }
        try { notificationListener = (NotificationCallback) getContext(); }
        catch (ClassCastException e) { throw new ClassCastException("Activity must implement NotificationListener"); }
    }

    /**
     * The Handler that gets information back from the BluetoothChatService
     */

    private final Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {

            switch (msg.what) {
                case Constants.MESSAGE_READ:
                    String output = (String) msg.obj;
                    if (!output.isEmpty()) {
                        readMessage(output);
                    }
                    break;
            }
        }
    };

    @Override
    public void onResume() {
        super.onResume();
        if (mChatService.getState() == Constants.STATE_CONNECTED) {
            checkVersionSent = true;
            writeToRPI("treehouses remote version " + BuildConfig.VERSION_CODE + "\n");
        }
    }
}

File: app/src/main/java/io/treehouses/remote/Fragments/HomeFragment.kt
package io.treehouses.remote.Fragments

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.app.ProgressDialog
import android.bluetooth.BluetoothAdapter
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.*
import android.preference.PreferenceManager
import android.view.ContextThemeWrapper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ExpandableListView
import android.widget.Toast
import androidx.core.content.ContextCompat
import io.treehouses.remote.*
import io.treehouses.remote.Constants.REQUEST_ENABLE_BT
import io.treehouses.remote.Fragments.DialogFragments.RPIDialogFragment
import io.treehouses.remote.InitialActivity.Companion.instance
import io.treehouses.remote.adapter.ProfilesListAdapter
import io.treehouses.remote.bases.BaseHomeFragment
import io.treehouses.remote.callback.NotificationCallback
import io.treehouses.remote.callback.SetDisconnect
import io.treehouses.remote.databinding.ActivityHomeFragmentBinding
import io.treehouses.remote.pojo.NetworkProfile
import io.treehouses.remote.utils.RESULTS
import io.treehouses.remote.utils.SaveUtils
import io.treehouses.remote.utils.match
import kotlinx.android.synthetic.main.activity_home_fragment.*
import java.util.*

class HomeFragment : BaseHomeFragment(), SetDisconnect {
    private var notificationListener: NotificationCallback? = null
    private var progressDialog: ProgressDialog? = null
    private var connectionState = false
    private var testConnectionResult = false
    private var testConnectionDialog: AlertDialog? = null
    private var selected_LED = 0
    private var checkVersionSent = false
    private var internetSent = false
    private var network_ssid = ""
    private var networkProfile: NetworkProfile? = null

    private lateinit var bind: ActivityHomeFragmentBinding
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        bind = ActivityHomeFragmentBinding.inflate(inflater, container, false)
        mChatService = listener.getChatService()
        mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        preferences = PreferenceManager.getDefaultSharedPreferences(context)
        setupProfiles()
        showDialogOnce(preferences!!)
        checkConnectionState()
        connectRpiListener()
        bind.btnGetStarted.setOnClickListener {
            instance!!.openCallFragment(AboutFragment())
            activity?.let { it.title = "About" }
        }
        testConnectionListener()
        return bind.root
    }

    private fun setupProfiles() {
        val profileAdapter = ProfilesListAdapter(context, listOf(*group_labels), SaveUtils.getProfiles(requireContext()))
        bind.networkProfiles.setAdapter(profileAdapter)
        bind.networkProfiles.setOnChildClickListener { _: ExpandableListView?, _: View?, groupPosition: Int, childPosition: Int, _: Long ->
            if (groupPosition == 3) {
                listener.sendMessage(getString(R.string.TREEHOUSES_DEFAULT_NETWORK))
                Toast.makeText(context, "Switched to Default Network", Toast.LENGTH_LONG).show()
            } else if (SaveUtils.getProfiles(requireContext()).size > 0 && SaveUtils.getProfiles(requireContext())[listOf(*group_labels)[groupPosition]]!!.size > 0) {
                if (SaveUtils.getProfiles(requireContext())[listOf(*group_labels)[groupPosition]]!!.size <= childPosition) return@setOnChildClickListener false
                networkProfile = SaveUtils.getProfiles(requireContext())[listOf(*group_labels)[groupPosition]]!![childPosition]
                listener.sendMessage(getString(R.string.TREEHOUSES_DEFAULT_NETWORK))
                Toast.makeText(requireContext(), "Configuring...", Toast.LENGTH_LONG).show()
            }
            false
        }
    }

    private fun switchProfile(profile: NetworkProfile?) {
        if (profile == null) return
        progressDialog = ProgressDialog.show(ContextThemeWrapper(context, R.style.CustomAlertDialogStyle), "Connecting...", "Switching to " + profile.ssid, true)
        progressDialog?.show()
        when {
            profile.isWifi -> {
                //WIFI
                listener.sendMessage(
                        getString(if (profile.isHidden) R.string.TREEHOUSES_WIFI_HIDDEN else R.string.TREEHOUSES_WIFI,
                        profile.ssid, profile.password))
                network_ssid = profile.ssid
            }
            profile.isHotspot -> {
                //Hotspot
                listener.sendMessage(
                        getString(if (profile.isHidden) R.string.TREEHOUSES_AP_HIDDEN else R.string.TREEHOUSES_AP,
                        profile.option, profile.ssid, profile.password))
                network_ssid = profile.ssid
            }
            profile.isBridge -> {
                //Bridge
                listener.sendMessage(getString(R.string.TREEHOUSES_BRIDGE, profile.ssid, profile.hotspot_ssid,
                        profile.password, profile.hotspot_password))
            }
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (MainApplication.showLogDialog) {
            rate(preferences!!)
            showLogDialog(preferences!!)
        }
    }

    private fun connectRpiListener() {
        bind.btnConnect.setOnClickListener {
            val vibe = requireContext().getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            if (vibe.hasVibrator()) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) vibe.vibrate(VibrationEffect.createOneShot(10, VibrationEffect.DEFAULT_AMPLITUDE))
                else vibe.vibrate(10)
            }
            if (connectionState) {
                RPIDialogFragment.getInstance().bluetoothCheck("unregister")
                mChatService.stop()
                connectionState = false
                checkConnectionState()
                return@setOnClickListener
            }
            if (mBluetoothAdapter?.state == BluetoothAdapter.STATE_OFF) {
                startActivityForResult(Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE), REQUEST_ENABLE_BT)
                Toast.makeText(context, "Bluetooth is disabled", Toast.LENGTH_LONG).show()
            } else if (mBluetoothAdapter?.state == BluetoothAdapter.STATE_ON) showRPIDialog(this@HomeFragment)
        }
    }

    private fun testConnectionListener() {
        bind.testConnection.setOnClickListener {
            val preference = androidx.preference.PreferenceManager.getDefaultSharedPreferences(context).getString("led_pattern", "LED Heavy Metal")
            val options = listOf(*resources.getStringArray(R.array.led_options))
            val optionsCode = resources.getStringArray(R.array.led_options_commands)
            selected_LED = options.indexOf(preference)
            listener.sendMessage(optionsCode[selected_LED])
            testConnectionDialog = showTestConnectionDialog(false, "Testing Connection...", R.string.test_connection_message, selected_LED)
            testConnectionDialog?.show()
            testConnectionResult = false
        }
    }

    override fun checkConnectionState() {
        mChatService = listener.getChatService()
        if (mChatService.state == Constants.STATE_CONNECTED) {
            showLogDialog(preferences!!)
            transition(true, arrayOf(150f, 110f, 70f))
            connectionState = true
            checkVersionSent = true
            listener.sendMessage(getString(R.string.TREEHOUSES_REMOTE_VERSION, BuildConfig.VERSION_CODE))
            Tutorials.homeTutorials(bind, requireActivity())
        } else {
            transition(false, arrayOf(0f, 0f, 0f))
            connectionState = false
            MainApplication.logSent = false
        }
        mChatService.updateHandler(mHandler)
    }

    private fun transition(connected: Boolean, values: Array<Float>) {
        bind.btnConnect.text = if (connected) "Disconnect" else "Connect to RPI"
        bind.btnConnect.setBackgroundResource(if (connected) R.drawable.ic_disconnect_rpi else R.drawable.ic_connect_to_rpi)
        bind.backgroundHome.animate().translationY(values[0])
        bind.btnConnect.animate().translationY(values[1])
        bind.btnGetStarted.animate().translationY(values[2])
        val b1 = if (connected) View.GONE else View.VISIBLE     //Show on Boot
        val b2 = if (connected) View.VISIBLE else View.GONE     //Show when connected
        bind.welcomeHome.visibility = b1
        bind.logoHome.visibility = b1
        bind.testConnection.visibility = b2
        bind.layoutBack.visibility = b2

    }

    private fun dismissTestConnection() {
        if (testConnectionDialog != null) {
            testConnectionDialog!!.cancel()
            showTestConnectionDialog(true, "Process Finished", R.string.test_finished, selected_LED)
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        notificationListener = try { getContext() as NotificationCallback?
        } catch (e: ClassCastException) {
            throw ClassCastException("Activity must implement NotificationListener")
        }
    }

    private fun dismissPDialog() {
        if (progressDialog != null) progressDialog!!.dismiss()
    }

    private fun checkVersion(output: String) {
        checkVersionSent = false
        if (output.contains("Usage") || output.contains("command")) {
            showUpgradeCLI()
        } else if (BuildConfig.VERSION_CODE == 2 || output.contains("true")) {
            listener.sendMessage(getString(R.string.TREEHOUSES_REMOTE_CHECK))
        } else if (output.contains("false")) {
            val alertDialog = AlertDialog.Builder(ContextThemeWrapper(context, R.style.CustomAlertDialogStyle))
                    .setTitle("Update Required")
                    .setMessage("Please update Treehouses Remote, as it does not meet the required version on the Treehouses CLI.")
                    .setPositiveButton("Update") { _: DialogInterface?, _: Int ->
                        val appPackageName = requireActivity().packageName // getPackageName() from Context or Activity object
                        try {
                            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$appPackageName")))
                        } catch (anfe: ActivityNotFoundException) {
                            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=$appPackageName")))
                        }
                    }.create()
            alertDialog.show()

        }
    }

    private fun readMessage(output: String) {
        notificationListener = try { context as NotificationCallback?
        } catch (e: ClassCastException) {
            throw ClassCastException("Activity must implement NotificationListener")
        }
        val s = match(output)
        when {
            s == RESULTS.ERROR && !output.toLowerCase(Locale.ROOT).contains("error") -> {
                showUpgradeCLI()
                internetSent = false
            }
            s == RESULTS.VERSION && checkVersionSent -> checkVersion(output)
            s == RESULTS.REMOTE_CHECK -> {
                checkImageInfo(output.trim().split(" "), mChatService.connectedDeviceName)
                listener.sendMessage(getString(R.string.TREEHOUSES_INTERNET))
                internetSent = true
            }
            s == RESULTS.BOOLEAN && internetSent -> {
                internetSent = false
                if (output.contains("true")) internetstatus?.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.circle_green))
                else internetstatus?.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.circle_green))
                listener.sendMessage(getString(R.string.TREEHOUSES_UPGRADE_CHECK))
            }
            else -> moreActions(output, s)
        }
    }

    private fun moreActions(output: String, result: RESULTS) {
        when {
            result == RESULTS.UPGRADE_CHECK -> notificationListener?.setNotification(output.contains("true"))
            result == RESULTS.HOTSPOT_CONNECTED || result == RESULTS.WIFI_CONNECTED -> {
                dismissPDialog()
                Toast.makeText(context, "Switched to $network_ssid", Toast.LENGTH_LONG).show()
            }
            result == RESULTS.BRIDGE_CONNECTED -> {
                dismissPDialog()
                Toast.makeText(context, "Bridge Has Been Built", Toast.LENGTH_LONG).show()
            }
            result == RESULTS.DEFAULT_NETWORK -> switchProfile(networkProfile)
            result == RESULTS.ERROR -> {
                dismissPDialog()
                Toast.makeText(context, "Network Not Found", Toast.LENGTH_LONG).show()
            }
            !testConnectionResult -> {
                testConnectionResult = true
                dismissTestConnection()
            }
        }
    }

    /**
     * The Handler that gets information back from the BluetoothChatService
     */
    private val mHandler: Handler = @SuppressLint("HandlerLeak")
    object : Handler() {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                Constants.MESSAGE_READ -> {
                    val output = msg.obj as String
                    if (output.isNotEmpty()) {
                        readMessage(output)
                    }
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (mChatService.state == Constants.STATE_CONNECTED) {
            checkVersionSent = true
            listener.sendMessage(getString(R.string.TREEHOUSES_REMOTE_VERSION, BuildConfig.VERSION_CODE))
        }
    }

    companion object {
        @JvmField
        val group_labels = arrayOf("WiFi", "Hotspot", "Bridge", "Default")
    }
}